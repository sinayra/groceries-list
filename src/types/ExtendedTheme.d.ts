export interface ExtendedTheme extends Theme {
    dark: boolean,
    colors: {
      primary: string,
      background: string,
      card: string,
      text: string,
      border: string,
      notification: string,
      red: string,
      green: string,
      yellow: string,
      secondaryCard: string,
    }
  }