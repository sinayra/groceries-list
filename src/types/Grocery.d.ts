export interface Purchase {
  id?: string;
  date: number;
  price: number;
  quantity: number;
}

export interface Grocery {
  listId?: string;
  listQuantity?: number,
  id?: string;
  name: string;
  purchases: Purchase[];
}
