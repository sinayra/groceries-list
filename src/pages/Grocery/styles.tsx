import { StyleSheet } from "react-native";
import Constants from "expo-constants";

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: "#fff",
    flexDirection: "column",
    paddingTop: 20 + Constants.statusBarHeight,
    paddingBottom: 20 + Constants.statusBarHeight,
    paddingLeft: 10,
    paddingRight: 10,
    justifyContent: "space-between",
    alignItems: "stretch",
    alignContent: "stretch",
  },
  content: {
    backgroundColor: "#FFA",
    flex: 1,
  },
  button: {
    backgroundColor: "#34CB79",
    height: 60,
    flexDirection: "row",
    borderRadius: 10,
    overflow: "hidden",
    alignItems: "center",
    marginTop: 8,
  },
  buttonText: {
    flex: 1,
    justifyContent: "center",
    textAlign: "center",
    color: "#FFFFFF",
    fontFamily: "Roboto_500Medium",
    fontSize: 16,
  },
  input: {
    height: 54,
    backgroundColor: "#FFF",
    justifyContent: "center",
    paddingHorizontal: 16,
    marginTop: 16,
    borderColor: '#FFF',
    borderBottomColor: '#CCC',
    borderWidth: 2,
  },
  label: {
    fontFamily: 'Roboto_400Regular',
    color: '#D4C2FF',
  },
  inputGroup: {
    flexDirection: 'row',
    justifyContent: 'space-between',
  },
  inputBlock: {
    width: '48%',
  },
  form: {
    flex: 1,
  },
  showNew: {
    flexDirection: 'row',
    alignItems: 'center'
  }
});

export default styles;
