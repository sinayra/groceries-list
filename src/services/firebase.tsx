import firebase from "firebase";
import {
  apiKey,
  authDomain,
  databaseURL,
  projectId,
  storageBucket,
  messagingSenderId,
  appId,
  measurementId,
} from "../config/keys.json";

const firebaseConfig = {
  apiKey,
  authDomain,
  databaseURL,
  projectId,
  storageBucket,
  messagingSenderId,
  appId,
  measurementId,
};

firebase.initializeApp(firebaseConfig);

export default firebase;
