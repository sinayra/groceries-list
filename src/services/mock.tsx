const mock = {
  "groceries" : {
    "0" : {
      "name" : "Coca 1L",
      "purchases" : [ {
        "date" : 1599264000000,
        "price" : 11.18,
        "quantity" : 2
      }, {
        "date" : 1599955200000,
        "price" : 10.38,
        "quantity" : 2
      }, {
        "date" : 1600041600000,
        "price" : 10.38,
        "quantity" : 2
      }, {
        "date" : 1600646400000,
        "price" : 11.18,
        "quantity" : 2
      } ]
    },
    "1" : {
      "name" : "Coca Zero 1.5L",
      "purchases" : [ {
        "date" : 1599264000000,
        "price" : 7.78,
        "quantity" : 2
      }, {
        "date" : 1600646400000,
        "price" : 3.89,
        "quantity" : 1
      } ]
    },
    "2" : {
      "name" : "Chá",
      "purchases" : [ {
        "date" : 1599264000000,
        "price" : 8.99,
        "quantity" : 1
      }, {
        "date" : 1600646400000,
        "price" : 17.98,
        "quantity" : 2
      } ]
    },
    "3" : {
      "name" : "Sabonete",
      "purchases" : [ {
        "date" : 1599264000000,
        "price" : 1.34,
        "quantity" : 1
      } ]
    },
    "4" : {
      "name" : "Batata",
      "purchases" : [ {
        "date" : 1594252800000,
        "price" : 2.77,
        "quantity" : 0.5551102204408818
      }, {
        "date" : 1595635200000,
        "price" : 1.4,
        "quantity" : 0.280561122244489
      }, {
        "date" : 1600041600000,
        "price" : 0.87,
        "quantity" : 0.2909698996655518
      }, {
        "date" : 1601078400000,
        "price" : 1.32,
        "quantity" : 0.44
      } ]
    },
    "5" : {
      "name" : "Cenoura",
      "purchases" : [ {
        "date" : 1595635200000,
        "price" : 1.0,
        "quantity" : 0.20040080160320642
      }, {
        "date" : 1595894400000,
        "price" : 0.81,
        "quantity" : 0.2903225806451613
      }, {
        "date" : 1600041600000,
        "price" : 0.74,
        "quantity" : 0.1544885177453027
      } ]
    },
    "6" : {
      "name" : "Pão de forma",
      "purchases" : [ {
        "date" : 1599264000000,
        "price" : 3.99,
        "quantity" : 1
      }, {
        "date" : 1600646400000,
        "price" : 3.19,
        "quantity" : 1
      }, {
        "date" : 1598313600000,
        "price" : 6.07,
        "quantity" : 1
      }, {
        "date" : 1593734400000,
        "price" : 4.39,
        "quantity" : 1
      }, {
        "date" : 1595116800000,
        "price" : 3.4,
        "quantity" : 1
      }, {
        "date" : 1591142400000,
        "price" : 4.19,
        "quantity" : 1
      }, {
        "date" : 1591401600000,
        "price" : 4.29,
        "quantity" : 1
      }, {
        "date" : 1592092800000,
        "price" : 4.79,
        "quantity" : 1
      }, {
        "date" : 1593043200000,
        "price" : 7.19,
        "quantity" : 1
      } ]
    },
    "-MJ7uyt7ZxyrE_XkrdQo" : {
      "name" : "mais um produto",
      "purchases" : {
        "-MJ7uyu6vHZ-hgRS8Lna" : {
          "date" : 1602172468142,
          "price" : 3,
          "quantity" : 3
        },
        "-MJ7v2o4EbJaTO90QG-T" : {
          "date" : 1602172468142,
          "price" : 2,
          "quantity" : 2
        }
      }
    }
  }
}

export default mock;
