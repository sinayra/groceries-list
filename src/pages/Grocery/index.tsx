import React from "react";
import {
  View,
  Text,
  TouchableOpacity,
  SafeAreaView,
  Linking,
} from "react-native";
import { RectButton } from "react-native-gesture-handler";
import styles from "./styles";

export default function Home() {
  function handleSavePurchase() {
    console.info("not implemented");
  }

  return (
    <SafeAreaView style={styles.container}>
      <View style={styles.content}>
        <View style={styles.form}>
          <Text>Grocery page</Text>
        </View>
        <RectButton style={styles.button} onPress={handleSavePurchase}>
          <Text style={styles.buttonText}>Salvar</Text>
        </RectButton>
      </View>
    </SafeAreaView>
  );
}
