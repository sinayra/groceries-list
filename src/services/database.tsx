import firebase from "./firebase";
import { Grocery, Purchase } from "../types/Grocery";

const database = firebase.database();

export async function getGroceries(): Promise<Grocery[]> {
  const promise: Promise<Grocery[]> = new Promise((resolve, reject) => {
    database.ref("/groceries").orderByChild("name").on("value", (snapshot) => {
      let groceries: Grocery[] = [];

      snapshot.forEach((elemSnapshot) => {
        let purchases: Purchase[] = [];
        const id = elemSnapshot.key ? elemSnapshot.key : undefined;
        const name = elemSnapshot.child("name").val();
        const obj = elemSnapshot.child("purchases");

        obj.forEach((purchasesSnapshot) => {
          const id = purchasesSnapshot.key ? purchasesSnapshot.key : undefined;
          const date = purchasesSnapshot.child("date").val();
          const price = purchasesSnapshot.child("price").val();
          const quantity = purchasesSnapshot.child("quantity").val();

          purchases.push({ id, date, price, quantity });
        });

        groceries.push({ id, name, purchases });
      });

      resolve(groceries);
    });
  });

  return promise;
}

export async function getPurchases(id: string): Promise<Purchase[]> {
  const promise: Promise<Purchase[]> = new Promise((resolve, reject) => {
    database.ref(`/groceries/${id}/purchases/`).on("value", (snapshot) => {
      let purchases: Purchase[] = [];

      snapshot.forEach((elemSnapshot) => {
        const id = elemSnapshot.key ? elemSnapshot.key : undefined;
        const date = elemSnapshot.child("date").val();
        const price = elemSnapshot.child("price").val();
        const quantity = elemSnapshot.child("quantity").val();

        purchases.push({ id, date, price, quantity });
      });

      resolve(purchases);
    });
  });

  return promise;
}

export async function setGroceries(
  id: string | undefined,
  name: string,
  date: number,
  price: number,
  quantity: number
) {
  let grocery: string | null | undefined = id;

  if (grocery === undefined) {
    grocery = (await database.ref("/groceries").push()).key;

    if (grocery !== null) {
      database.ref(`groceries/${grocery}`).set({
        name,
        purchases: {},
      });
    }
  }

  if (grocery !== null) {
    const newPurchase = (await database.ref(`groceries/${grocery}`).push()).key;

    database.ref(`groceries/${grocery}/purchases/${newPurchase}`).set({
      date,
      price,
      quantity,
    });

    return 200;
  } else {
    return 500;
  }
}

export async function deleteGroceries(
  idGrocery: string,
  idPurchase: string | null = null
) {
  if (idPurchase === null) {
    await database.ref(`/groceries/${idGrocery}`).remove();
  } else {
    await database
      .ref(`/groceries/${idGrocery}/purchases/${idPurchase}`)
      .remove();
  }

  return 200;
}

export async function getPurchaseList(): Promise<Grocery[]> {
  const promise: Promise<Grocery[]> = new Promise((resolve, reject) => {
    database.ref("/list").on("value", (snapshot) => {
      let groceries: Grocery[] = [];
      console.log(snapshot);

      snapshot.forEach((listId) => {

        const id = listId.key ? listId.key : undefined;
        const groceryId = listId.child("id").val();
        console.log("ID", id)

        database.ref(`/groceries/${groceryId}`).on("value", (grocery) => {
          console.log("GROCERY", grocery)
          let purchases: Purchase[] = [];
          const name = grocery.child("name").val();
          const obj = grocery.child("purchases");

          obj.forEach((purchasesSnapshot) => {
            const purchaseId = purchasesSnapshot.key ? purchasesSnapshot.key : undefined;
            const date = purchasesSnapshot.child("date").val();
            const price = purchasesSnapshot.child("price").val();
            const quantity = purchasesSnapshot.child("quantity").val();

            purchases.push({ id: purchaseId, date, price, quantity });
          });

          groceries.push({ listId: id, id: groceryId, name, purchases });
        });


      });

      resolve(groceries);
    });

  });

  return promise;
}

export async function addToPurchaseList(
  id: string | undefined,
) {
  if (id) {
    const key = (await database.ref(`/list/${id}`).push()).key;
    database.ref(`list/${key}`).set({
      id
    });
    return 200;
  } else {
    return 500;
  }
}

export async function removeFromPurchaseList(
  id: string | undefined,
) {
  if (id) {
    await database.ref(`/list/${id}`).remove();

    return 200;
  } else {
    return 500;
  }
}
