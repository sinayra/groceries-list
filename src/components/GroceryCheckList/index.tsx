import React, { useEffect, useState } from "react";
import { FlatList, Text, TouchableHighlight, ToastAndroid, View } from "react-native";

import { addToPurchaseList, insertGrocery } from "../../services/database";
import { Grocery } from "../../types/Grocery";
import GroceryCheckItem from "../GroceryCheckItem";
import { Variables } from "../../styles/variables";

interface GroceryCheckListProps {
    showList: Grocery[];
    purchaseList?: Grocery[];
    canBeAddedToList: boolean,
    reload: () => void;
    filter?: string
}

const GroceryCheckList: React.FC<GroceryCheckListProps> = ({ showList, purchaseList, canBeAddedToList, reload, filter }) => {
    const variables = Variables();
    const [data, setData] = useState<Grocery[]>(showList);
    const [emptyMessageOption, setEmptyMessageOption] = useState<"EXISTE" | "ADICIONAR" | "NAO EXISTE">();
    const [emptyMessage, setEmptyMessage] = useState();

    async function handleAddGrocery() {
        if (filter) {
            const grocery = await insertGrocery(filter);
            if (grocery) {
                addToPurchaseList(grocery);

                reload();
            }
            else {
                ToastAndroid.show(`Erro ao salvar no Firebase`, ToastAndroid.SHORT);
            }
        }
        else {
            ToastAndroid.show(`Não é possível adicionar ${filter} na Lista de Compra`, ToastAndroid.SHORT);
        }
    }

    function handleEmptyMessage() {
        if (canBeAddedToList && purchaseList && filter) {
            const isInPurchaseList = purchaseList.find((elem) => elem.name === filter);

            if (isInPurchaseList) {
                setEmptyMessageOption("EXISTE");
            }
            else {
                setEmptyMessageOption("ADICIONAR");
            }
        }
        else {
            setEmptyMessageOption("NAO EXISTE");
        }

    }

    useEffect(() => {
        if (purchaseList) {
            const result = showList.filter((elem) => {
                for (let i = 0; i < purchaseList.length; i++) {
                    if (elem.id === purchaseList[i].id) {
                        return false;
                    }
                }
                return true;
            });
            setData(result);
        }
        else {
            setData(showList);
        }

        handleEmptyMessage();
    }, [showList, purchaseList]);


    function displayEmptyMessage() {
        switch (emptyMessageOption) {
            case "ADICIONAR":
                return (<TouchableHighlight onPress={handleAddGrocery}>
                    <View style={{ flexDirection: "row" }}>
                        <Text style={{ color: variables.PRIMARY_COLOR, fontSize: variables.FONT_SIZE_MEDIUM }}>Gostaria de adicionar </Text>
                        <Text style={{ color: variables.PRIMARY_COLOR, fontSize: variables.FONT_SIZE_MEDIUM, fontWeight: "bold" }}>{filter}</Text>
                        <Text style={{ color: variables.PRIMARY_COLOR, fontSize: variables.FONT_SIZE_MEDIUM }}> ?</Text>
                    </View>
                </TouchableHighlight>);

            case "EXISTE":
                return (<View style={{ flexDirection: "row" }}>
                    <Text style={{ color: variables.TEXT_COLOR, fontSize: variables.FONT_SIZE_MEDIUM, fontWeight: "bold" }}>{filter}</Text>
                    <Text style={{ color: variables.TEXT_COLOR, fontSize: variables.FONT_SIZE_MEDIUM }}> já está na Lista de Compras.</Text>
                </View>);

            default:
                return (<Text style={{ color: variables.TEXT_COLOR, fontSize: variables.FONT_SIZE_MEDIUM }}>Não há itens na sua Lista.</Text>)

        }

    };

    return (
        <>
            {
                data.length > 0 ?
                    <FlatList
                        data={data}
                        renderItem={({ item, index }: { item: Grocery, index: number }) => <GroceryCheckItem item={item} canBeAddedToList={canBeAddedToList} reload={reload} />}
                        keyExtractor={(item) => item.id as string}
                    />
                    :
                    <>
                        {displayEmptyMessage()}
                    </>
            }

        </>
    );
};

export default GroceryCheckList;
