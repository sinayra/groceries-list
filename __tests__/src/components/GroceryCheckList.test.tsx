import React from 'react';
import { shallow } from 'enzyme';
import { Text, ToastAndroid } from 'react-native';

import GroceryCheckList from "../../../src/components/GroceryCheckList";
import { Grocery } from "../../../src/types/Grocery";
import * as database from "../../../src/services/database";

describe('<Grocery Check List />', () => {
    let mockAddToPurchaseList: jest.SpyInstance<Promise<200 | 500>, [id: string | undefined]>;
    let mockInsertGrocery: jest.SpyInstance<Promise<string | null>, [name: string]>;
    let mockToastAndroid: jest.SpyInstance<void | null, [message: string, duration: number]>;

    beforeEach(() => {
        mockAddToPurchaseList = jest.spyOn(database, 'addToPurchaseList').mockImplementation(jest.fn());
        mockInsertGrocery = jest.spyOn(database, 'insertGrocery').mockImplementation(() => {
            const promise = new Promise<string | null>((resolve, reject) => {
                resolve("foo");
            });

            return promise;
        });

        mockToastAndroid = jest.spyOn(ToastAndroid, 'show').mockImplementation(jest.fn());
    });

    afterEach(() => {
        mockAddToPurchaseList.mockRestore();
        mockInsertGrocery.mockRestore();
        mockToastAndroid.mockRestore();
    });

    describe('Items already in purchase list', () => {
        let reload: () => void;
        let addToList: boolean;

        beforeAll(() => {
            reload = jest.fn();
            addToList = false;
        });

        describe('loads without crashing', () => {
            it('Snapshot', async () => {
                const showList: Grocery[] = [{
                    name: "Sabonete",
                    listId: "0",
                    listQuantity: 1,
                    purchases: [{
                        "date": 1599264000000,
                        "price": 1.34,
                        "quantity": 1
                    }]
                }];

                const component = shallow(<GroceryCheckList showList={showList} canBeAddedToList={addToList} reload={reload} />);

                expect(component.getElement()).toMatchSnapshot();
            });

            it('multiple elements', () => {
                const showList: Grocery[] = [
                    {
                        name: "Sabonete",
                        listId: "0",
                        listQuantity: 1,
                        purchases: [{
                            "date": 1599264000000,
                            "price": 1.34,
                            "quantity": 1
                        }]
                    },
                    {
                        listId: "1",
                        listQuantity: 3,
                        name: "Batata",
                        purchases: [{
                            "date": 1594252800000,
                            "price": 2.77,
                            "quantity": 0.5551102204408818
                        }, {
                            "date": 1595635200000,
                            "price": 1.4,
                            "quantity": 0.280561122244489
                        }, {
                            "date": 1600041600000,
                            "price": 0.87,
                            "quantity": 0.2909698996655518
                        }, {
                            "date": 1601078400000,
                            "price": 1.32,
                            "quantity": 0.44
                        }]
                    },
                    {
                        listId: "2",
                        listQuantity: 1,
                        name: "Chá",
                        purchases: [{
                            "date": 1599264000000,
                            "price": 8.99,
                            "quantity": 1
                        }, {
                            "date": 1600646400000,
                            "price": 17.98,
                            "quantity": 2
                        }]
                    }
                ];

                const component = shallow(<GroceryCheckList showList={showList} canBeAddedToList={addToList} reload={reload} />);
                const list = component.find({ testID: "list" });

                expect(list).toHaveLength(1);
            });

            it('one element', () => {
                const showList: Grocery[] = [{
                    name: "Sabonete",
                    listId: "0",
                    listQuantity: 1,
                    purchases: [{
                        "date": 1599264000000,
                        "price": 1.34,
                        "quantity": 1
                    }]
                }];

                const component = shallow(<GroceryCheckList showList={showList} canBeAddedToList={addToList} reload={reload} />);
                const list = component.find({ testID: "list" });

                expect(list).toHaveLength(1);
            });

            it('no elements', () => {
                const showList: Grocery[] = []

                const component = shallow(<GroceryCheckList showList={showList} canBeAddedToList={addToList} reload={reload} />);
                const list = component.find({ testID: "list" });

                expect(list).toHaveLength(0);
            });
        });

    });

    describe('Items that can be added to purchase list', () => {
        let reload: () => void;
        let addToList: boolean;

        beforeAll(() => {
            reload = jest.fn();
            addToList = true;
        });

        describe('loads without crashing', () => {

            it('Snapshot', async () => {
                const showList: Grocery[] = [{
                    name: "Sabonete",
                    purchases: [{
                        "date": 1599264000000,
                        "price": 1.34,
                        "quantity": 1
                    }]
                }];

                const component = shallow(<GroceryCheckList showList={showList} canBeAddedToList={addToList} reload={reload} />);

                expect(component.getElement()).toMatchSnapshot();
            });

            it('multiple elements', () => {
                const showList: Grocery[] = [
                    {
                        name: "Sabonete",
                        purchases: [{
                            "date": 1599264000000,
                            "price": 1.34,
                            "quantity": 1
                        }]
                    },
                    {
                        name: "Batata",
                        purchases: [{
                            "date": 1594252800000,
                            "price": 2.77,
                            "quantity": 0.5551102204408818
                        }, {
                            "date": 1595635200000,
                            "price": 1.4,
                            "quantity": 0.280561122244489
                        }, {
                            "date": 1600041600000,
                            "price": 0.87,
                            "quantity": 0.2909698996655518
                        }, {
                            "date": 1601078400000,
                            "price": 1.32,
                            "quantity": 0.44
                        }]
                    },
                    {
                        name: "Chá",
                        purchases: [{
                            "date": 1599264000000,
                            "price": 8.99,
                            "quantity": 1
                        }, {
                            "date": 1600646400000,
                            "price": 17.98,
                            "quantity": 2
                        }]
                    }
                ];

                const component = shallow(<GroceryCheckList showList={showList} canBeAddedToList={addToList} reload={reload} />);
                const list = component.find({ testID: "list" });

                expect(list).toHaveLength(1);
            });

            it('one element', () => {
                const showList: Grocery[] = [{
                    name: "Sabonete",
                    purchases: [{
                        "date": 1599264000000,
                        "price": 1.34,
                        "quantity": 1
                    }]
                }];

                const component = shallow(<GroceryCheckList showList={showList} canBeAddedToList={addToList} reload={reload} />);
                const list = component.find({ testID: "list" });

                expect(list).toHaveLength(1);
            });

            it('no elements', () => {
                const showList: Grocery[] = []

                const component = shallow(<GroceryCheckList showList={showList} canBeAddedToList={addToList} reload={reload} />);
                const list = component.find({ testID: "list" });
                const text = component.find(Text).getElement().props.children;

                expect(list).toHaveLength(0);
                expect(text).toBe("Não há itens na sua Lista.");
            });

            describe('has items in purchase list', () => {
                it('has elements to render', () => {
                    const showList: Grocery[] = [
                        {
                            id: "1",
                            name: "Sabonete",
                            listId: "0",
                            listQuantity: 1,
                            purchases: [{
                                "date": 1599264000000,
                                "price": 1.34,
                                "quantity": 1
                            }]
                        },
                        {
                            id: "2",
                            name: "Batata",
                            purchases: [{
                                "date": 1594252800000,
                                "price": 2.77,
                                "quantity": 0.5551102204408818
                            }, {
                                "date": 1595635200000,
                                "price": 1.4,
                                "quantity": 0.280561122244489
                            }, {
                                "date": 1600041600000,
                                "price": 0.87,
                                "quantity": 0.2909698996655518
                            }, {
                                "date": 1601078400000,
                                "price": 1.32,
                                "quantity": 0.44
                            }]
                        },
                        {
                            id: "3",
                            name: "Chá",
                            purchases: [{
                                "date": 1599264000000,
                                "price": 8.99,
                                "quantity": 1
                            }, {
                                "date": 1600646400000,
                                "price": 17.98,
                                "quantity": 2
                            }]
                        }
                    ];

                    const purchaseList: Grocery[] = [
                        {
                            id: "1",
                            name: "Sabonete",
                            listId: "0",
                            listQuantity: 1,
                            purchases: [{
                                "date": 1599264000000,
                                "price": 1.34,
                                "quantity": 1
                            }]
                        }
                    ];

                    const component = shallow(<GroceryCheckList showList={showList} canBeAddedToList={addToList} reload={reload} purchaseList={purchaseList} />);
                    const list = component.find({ testID: "list" });

                    expect(list).toHaveLength(1);
                });

                it('does not have elements to render', () => {
                    const showList: Grocery[] = [
                        {
                            id: "1",
                            name: "Sabonete",
                            listId: "0",
                            listQuantity: 1,
                            purchases: [{
                                "date": 1599264000000,
                                "price": 1.34,
                                "quantity": 1
                            }]
                        },
                        {
                            id: "2",
                            name: "Batata",
                            purchases: [{
                                "date": 1594252800000,
                                "price": 2.77,
                                "quantity": 0.5551102204408818
                            }, {
                                "date": 1595635200000,
                                "price": 1.4,
                                "quantity": 0.280561122244489
                            }, {
                                "date": 1600041600000,
                                "price": 0.87,
                                "quantity": 0.2909698996655518
                            }, {
                                "date": 1601078400000,
                                "price": 1.32,
                                "quantity": 0.44
                            }]
                        },
                        {
                            id: "3",
                            name: "Chá",
                            purchases: [{
                                "date": 1599264000000,
                                "price": 8.99,
                                "quantity": 1
                            }, {
                                "date": 1600646400000,
                                "price": 17.98,
                                "quantity": 2
                            }]
                        }
                    ];

                    const purchaseList: Grocery[] = [
                        {
                            id: "1",
                            name: "Sabonete",
                            listId: "0",
                            listQuantity: 1,
                            purchases: [{
                                "date": 1599264000000,
                                "price": 1.34,
                                "quantity": 1
                            }]
                        },
                        {
                            id: "2",
                            name: "Batata",
                            purchases: [{
                                "date": 1594252800000,
                                "price": 2.77,
                                "quantity": 0.5551102204408818
                            }, {
                                "date": 1595635200000,
                                "price": 1.4,
                                "quantity": 0.280561122244489
                            }, {
                                "date": 1600041600000,
                                "price": 0.87,
                                "quantity": 0.2909698996655518
                            }, {
                                "date": 1601078400000,
                                "price": 1.32,
                                "quantity": 0.44
                            }]
                        },
                        {
                            id: "3",
                            name: "Chá",
                            purchases: [{
                                "date": 1599264000000,
                                "price": 8.99,
                                "quantity": 1
                            }, {
                                "date": 1600646400000,
                                "price": 17.98,
                                "quantity": 2
                            }]
                        }
                    ];

                    const component = shallow(<GroceryCheckList showList={showList} canBeAddedToList={addToList} reload={reload} purchaseList={purchaseList} />);
                    const list = component.find({ testID: "list" });
                    const text = component.find(Text).getElement().props.children;

                    expect(list).toHaveLength(0);
                    expect(text).toBe("Não há itens na sua Lista.");
                });

                describe('search item by name', () => {
                    it('already in purchase list', () => {
                        const showList: Grocery[] = [
                            {
                                id: "1",
                                name: "Sabonete",
                                listId: "0",
                                listQuantity: 1,
                                purchases: [{
                                    "date": 1599264000000,
                                    "price": 1.34,
                                    "quantity": 1
                                }]
                            },
                        ];

                        const purchaseList: Grocery[] = [
                            {
                                id: "1",
                                name: "Sabonete",
                                listId: "0",
                                listQuantity: 1,
                                purchases: [{
                                    "date": 1599264000000,
                                    "price": 1.34,
                                    "quantity": 1
                                }]
                            }
                        ];

                        const filter = "Sabonete";

                        const component = shallow(<GroceryCheckList showList={showList} canBeAddedToList={addToList} reload={reload} purchaseList={purchaseList} filter={filter} />);

                        const list = component.find({ testID: "list" });
                        const text = component.find(Text).getElements();

                        const name = text[0].props.children;
                        const message = text[1].props.children;

                        expect(list).toHaveLength(0);
                        expect(name).toBe(filter);
                        expect(message).toBe(" já está na Lista de Compras.");

                    });

                    describe('not in purchase list', () => {
                        it('loads text', () => {
                            const showList: Grocery[] = [];
                            const purchaseList: Grocery[] = [];
                            const filter = "Sabonete";

                            const component = shallow(<GroceryCheckList showList={showList} canBeAddedToList={addToList} reload={reload} purchaseList={purchaseList} filter={filter} />);

                            const list = component.find({ testID: "list" });
                            const text = component.find(Text).getElements();

                            const name = text[1].props.children;
                            const message1 = text[0].props.children;
                            const message2 = text[2].props.children;

                            expect(list).toHaveLength(0);
                            expect(name).toBe(filter);
                            expect(message1).toBe("Gostaria de adicionar ");
                            expect(message2).toBe(" ?");

                        });

                        it('add new item', async () => {
                            const showList: Grocery[] = [];
                            const purchaseList: Grocery[] = [];
                            const filter = "Sabonete";

                            const component = shallow(<GroceryCheckList showList={showList} canBeAddedToList={addToList} reload={reload} purchaseList={purchaseList} filter={filter} />);

                            const add = component.find({ testID: "add" });
                            const handler = add.prop("onPress");

                            if (handler !== undefined) {
                                handler(true);
                            }

                            await expect(mockInsertGrocery).toBeCalled();
                            expect(mockAddToPurchaseList).toBeCalled();
                            expect(mockToastAndroid).not.toBeCalled();

                        });

                        describe('handle erros', () => {
                            it('database error', async () => {
                                const showList: Grocery[] = [];
                                const purchaseList: Grocery[] = [];
                                const filter = "Sabonete";

                                mockInsertGrocery.mockResolvedValueOnce(null);

                                const component = shallow(<GroceryCheckList showList={showList} canBeAddedToList={addToList} reload={reload} purchaseList={purchaseList} filter={filter} />);

                                const add = component.find({ testID: "add" });
                                const handler = add.prop("onPress");

                                if (handler !== undefined) {
                                    handler();
                                }

                                await expect(mockInsertGrocery).toBeCalled();
                                expect(mockAddToPurchaseList).not.toBeCalled();

                            });

                            it('filter error', async () => {
                                const showList: Grocery[] = [];
                                const purchaseList: Grocery[] = [];
                                const filter = "";

                                const component = shallow(<GroceryCheckList showList={showList} canBeAddedToList={addToList} reload={reload} purchaseList={purchaseList} filter={filter} />);

                                const add = component.find({ testID: "add" });
                                const handler = add.prop("onPress");

                                if (handler !== undefined) {
                                    handler();
                                }

                                await expect(mockInsertGrocery).not.toBeCalled();
                                expect(mockAddToPurchaseList).not.toBeCalled();

                            });
                        });
                    })
                });
            });
        });
    });
});