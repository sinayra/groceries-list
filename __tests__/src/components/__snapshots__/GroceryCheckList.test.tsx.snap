// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Grocery Check List /> Items already in purchase list loads without crashing Snapshot 1`] = `
<React.Fragment>
  <FlatList
    data={
      Array [
        Object {
          "listId": "0",
          "listQuantity": 1,
          "name": "Sabonete",
          "purchases": Array [
            Object {
              "date": 1599264000000,
              "price": 1.34,
              "quantity": 1,
            },
          ],
        },
      ]
    }
    disableVirtualization={false}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onEndReachedThreshold={2}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    testID="list"
    updateCellsBatchingPeriod={1000}
    windowSize={21}
  />
</React.Fragment>
`;

exports[`<Grocery Check List /> Items that can be added to purchase list loads without crashing Snapshot 1`] = `
<React.Fragment>
  <FlatList
    data={
      Array [
        Object {
          "name": "Sabonete",
          "purchases": Array [
            Object {
              "date": 1599264000000,
              "price": 1.34,
              "quantity": 1,
            },
          ],
        },
      ]
    }
    disableVirtualization={false}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onEndReachedThreshold={2}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    testID="list"
    updateCellsBatchingPeriod={1000}
    windowSize={21}
  />
</React.Fragment>
`;
